In [1]:
import warnings;
warnings.filterwarnings('ignore');

import pandas as pd

For this exam you are asked to complete the Python code between the 

```
#start code here

#end code here
```

comments. A direct question needs to be answered in the text field just below the question. 

You are allowed to use all the information available on the internet.

**question 1 (2pts)**

Create a Pandas DataFrame called `patients` that looks like this:

|           | gene_1 | gene_2 |
|-----------|--------|--------|
| patient_1 | 0.54   | 0.32   |
| patient_2 | 0.66   | 0.10   |

In [2]:
#start code here

patients = pd.DataFrame([[0.54,0.32],[0.66,0.1]],columns=["gene_1","gene_2"],index=["patient_1","patient_2"])

#end code here

patients

gene_1  gene_2
patient_1    0.54    0.32
patient_2    0.66    0.10

**question 2 (1pt)**

Use the Pandas DataFrame `.head()` function to print the first row of the `patients` DataFrame.

In [3]:
#start code here

print(patients.head(1))

#end code here

           gene_1  gene_2
patient_1    0.54    0.32


The following code loads the dataset we will use for this exam.

In [26]:
from sklearn.datasets import load_breast_cancer
dataset = load_breast_cancer(as_frame=True).frame
dataset.rename(columns = {'target':'label'}, inplace = True)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

**question 3 (1pt)**
    
Print the number of rows and columns in the `dataset` DataFrame.

In [5]:
#start code here

print("Number of rows = {}".format(dataset.shape[0]))
print("Number of columns = {}".format(dataset.shape[1]))

#end code here

Number of rows = 569
Number of columns = 31


**question 4 (1pt)**

Print the column `label` from the `dataset` DataFrame. Don't use `.loc` or `.iloc`.

In [6]:
#start code here

print(dataset.label)

#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64


**question 5 (1pt)**

Print the column `label` from the `dataset` DataFrame using `.loc`.

In [7]:
#start code here

print(dataset.loc[:,'label'])

#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64


**question 6 (1pt)**

Print the column `label` from the `dataset` DataFrame using `.iloc`.

In [8]:
#start code here

print(dataset.iloc[:,-1])

#end code here

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64


**question 7 (1pt)**

Print the label of the third row in the `dataset` DataFrame.

In [9]:
#start code here

print(dataset.label[2])
#or
print(dataset.loc[2,'label'])
#or
print(dataset.iloc[2,-1])

#end code here

0
0
0


**question 8 (1pt)**

Print the subset of the `dataset` DataFrame that only contains the columns in the Python list `cols_to_print`. You have to use the `cols_to_print` list in the code that creates the subset.

In [10]:
cols_to_print = ['mean area','mean concavity','mean symmetry','label']

#start code here

dataset.loc[:,cols_to_print]

#end code here

mean area  mean concavity  mean symmetry  label
0       1001.0         0.30010         0.2419      0
1       1326.0         0.08690         0.1812      0
2       1203.0         0.19740         0.2069      0
3        386.1         0.24140         0.2597      0
4       1297.0         0.19800         0.1809      0
..         ...             ...            ...    ...
564     1479.0         0.24390         0.1726      0
565     1261.0         0.14400         0.1752      0
566      858.1         0.09251         0.1590      0
567     1265.0         0.35140         0.2397      0
568      181.0         0.00000         0.1587      1

[569 rows x 4 columns]

**question 9 (2pts)**

Print all the rows in the `dataset` DataFrame with `mean symmetry` > 0.2 and `label` equal to 1.

In [11]:
#start code here

dataset.loc[(dataset['mean symmetry'] > 0.2) & (dataset['label'] == 1)]

#end code here

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
60        10.170         14.88           64.55      311.9          0.11340   
63         9.173         13.86           59.20      260.9          0.07721   
68         9.029         17.33           58.79      250.5          0.10660   
76        13.530         10.94           87.91      559.2          0.12910   
84        12.000         15.65           76.95      443.3          0.09723   
89        14.640         15.24           95.77      651.9          0.11320   
104       10.490         19.29           67.41      336.1          0.09989   
128       15.100         16.39           99.58      674.5          0.11500   
150       13.000         20.78           83.51      519.4          0.11350   
151        8.219         20.70           53.27      203.9          0.09405   
152        9.731         15.34           63.78      300.2          0.10720   
225       14.340         13.47           92.51      641.2          0.09906   
242       11.300         18.19           73.93      389.4          0.09592   
268       12.870         16.21           82.38      512.2          0.09425   
273        9.742         15.67           61.50      289.9          0.09037   
275       11.890         17.36           76.20      435.6          0.12250   
279       13.850         15.18           88.99      587.4          0.09516   
281       11.740         14.02           74.24      427.3          0.07813   
288       11.260         19.96           73.72      394.1          0.08020   
289       11.370         18.89           72.17      396.0          0.08713   
314        8.597         18.60           54.09      221.2          0.10740   
318        9.042         18.90           60.07      244.5          0.09968   
341        9.606         16.84           61.64      280.5          0.08481   
344       11.710         15.45           75.03      420.3          0.11500   
349       11.950         14.96           77.23      426.7          0.11580   
356       13.050         18.59           85.09      512.0          0.10820   
376       10.570         20.22           70.15      338.3          0.09073   
380       11.270         12.96           73.16      386.3          0.12370   
381       11.040         14.93           70.67      372.7          0.07987   
415       11.890         21.17           76.39      433.8          0.09773   
416        9.405         21.70           59.60      271.2          0.10440   
420       11.570         19.04           74.20      409.7          0.08546   
421       14.690         13.98           98.22      656.1          0.10310   
424        9.742         19.12           61.93      289.7          0.10750   
427       10.800         21.98           68.79      359.9          0.08801   
443       10.570         18.32           66.82      340.9          0.08142   
470        9.667         18.49           61.49      289.1          0.08946   
485       12.450         16.41           82.85      476.7          0.09514   
504        9.268         12.87           61.49      248.7          0.16340   
505        9.676         13.14           64.12      272.5          0.12550   
506       12.220         20.04           79.47      453.1          0.10960   
519       12.750         16.70           82.51      493.8          0.11250   
520        9.295         13.90           59.96      257.8          0.13710   
537       11.690         24.44           76.37      406.4          0.12360   
539        7.691         25.44           48.34      170.4          0.08668   
551       11.130         22.44           71.49      378.4          0.09566   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
60            0.08061        0.010840              0.01290         0.2743   
63            0.08751        0.059880              0.02180         0.2341   
68            0.14130        0.313000              0.04375         0.2111   
76            0.10470        0

**question 10 (2pts)**

Print the median of the `area error` column in `dataset` for all rows with `label` equal to 1.

In [12]:
#start code here

print(dataset.loc[dataset.label==1]["area error"].mean())

#end code here

21.135148459383736


**question 11 (2pts)**

Print the value of the `worst area` column in `dataset` for the row with the highest value for the `worst compactness` column.

In [13]:
#start code here

worst_idx = dataset["worst compactness"].idxmax()
dataset.loc[worst_idx,'worst area']

#end code here

711.4

**question 12 (2pts)**

Create a Pandes Series called `mean_area_ordinal` that, for each row in `dataset`, contains the value `low` if the value for `mean area` is within [0,420[, the value `middle` if the value for `mean area` is within [420,550[ and the value `high` if the value for `mean area` is larger of equal to 550. Use the `.apply()` function.

In [14]:
#start code here

def make_ordinal(x):
    if x["mean area"] < 420:
        return "low"
    elif x["mean area"] < 550:
        return "middle"
    else:
        return "high"
    
mean_area_ordinal = dataset.apply(make_ordinal,axis="columns")

#end code here

print(mean_area_ordinal)

0      high
1      high
2      high
3       low
4      high
       ... 
564    high
565    high
566    high
567    high
568     low
Length: 569, dtype: object


**question 13 (1pt)**

Write one line of code that shows how many datapoints in `mean_area_ordinal` have value `high`, how many have value `middle` and how many have value `low`.

In [15]:
#start code here

mean_area_ordinal.value_counts()

#end code here

high      285
low       142
middle    142
dtype: int64

**question 14 (2pts)**

To prepare for Scikit-learn model fitting we need to keep the feature vectors and the label separate. Create a variable `X` that contains all the 30 features in `dataset` and a variable `y` that contains the label.

In [16]:
#start code here

X = dataset.iloc[:,:-1]
y = dataset.iloc[:,-1]

#start code here

print(X.shape)
print(y.shape)

(569, 30)
(569,)


**question 15 (1pt)**

Use the Scikit-learn `StandardScaler` to standardize the feature values in `X` (assign the result to `X_norm`).

In [17]:
from sklearn.preprocessing import StandardScaler

#start code here

X_norm = StandardScaler().fit_transform(X)

#start code here

print(X_norm)

[[ 1.09706398 -2.07333501  1.26993369 ...  2.29607613  2.75062224
   1.93701461]
 [ 1.82982061 -0.35363241  1.68595471 ...  1.0870843  -0.24388967
   0.28118999]
 [ 1.57988811  0.45618695  1.56650313 ...  1.95500035  1.152255
   0.20139121]
 ...
 [ 0.70228425  2.0455738   0.67267578 ...  0.41406869 -1.10454895
  -0.31840916]
 [ 1.83834103  2.33645719  1.98252415 ...  2.28998549  1.91908301
   2.21963528]
 [-1.80840125  1.22179204 -1.81438851 ... -1.74506282 -0.04813821
  -0.75120669]]


**question 16 (1pt)**

What does the following code do? What is in `X_train`, `X_test`, `y_train` and `y_test`?

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.2, random_state=0)

*Answer here*

It splits the data *at random* into a training set (feature vectors in X_train and label in y_train (80% of the data)) and a test set (feature vectors in X_test and label in y_test (20% of the data))

**question 17 (2pts)**

Use the training set to compute the opimal values for the modelparameters in a Logistic Regression model (with hyperparameter `random_state` set to 1, all other hyperparameter values should be set to their Scikit-learn default values).

In [19]:
from sklearn.linear_model import LogisticRegression

#start code here

lr_model = LogisticRegression(random_state=1)
lr_model.fit(X_train,y_train)

#end code here

LogisticRegression(random_state=1)

**question 18 (2pts)**

Given the fitted model, compute predictions on the test set and compute the Area Under the Receiver Operating Characteristic curve (AUC) for these predictions.

In [20]:
from sklearn.metrics import roc_auc_score

#start code here

predictions = pd.DataFrame(lr_model.predict_proba(X_test),columns=["prob_0","prob_1"])

score_auc = roc_auc_score(y_test,predictions["prob_1"])

#end code here

print(score_auc)

0.9933312162591299


**question 19 (2pts)**

Create a Pandas DataFrame `df_model_parameters` with a column `feature` that contains the feature names in `X` and a column `parameter_value` that contains the corresponding fitted modelparameter value.

In [21]:
#start code here

df_model_parameters = pd.DataFrame()
df_model_parameters["feature"] = X.columns
df_model_parameters["parameter_value"] = lr_model.coef_[0]

#end code here

print(df_model_parameters)

                    feature  parameter_value
0               mean radius        -0.325119
1              mean texture        -0.494943
2            mean perimeter        -0.317433
3                 mean area        -0.395679
4           mean smoothness        -0.188723
5          mean compactness         0.436253
6            mean concavity        -0.676390
7       mean concave points        -0.843549
8             mean symmetry        -0.334403
9    mean fractal dimension         0.209551
10             radius error        -1.376258
11            texture error         0.045093
12          perimeter error        -0.812926
13               area error        -0.994695
14         smoothness error         0.257041
15        compactness error         0.671800
16          concavity error         0.102279
17     concave points error        -0.225080
18           symmetry error         0.141223
19  fractal dimension error         0.872027
20             worst radius        -0.922528
21        

**question 20 (1pt)**

Add a comlumn `parameter_value_abs` to DataFrame `df_model_parameters` that contains the absolute value of the modelparameter values in column `parameter_value`.

In [22]:
#start code here

df_model_parameters["parameter_value_abs"] = df_model_parameters["parameter_value"].map(abs)

#end code here

print(df_model_parameters)

                    feature  parameter_value  parameter_value_abs
0               mean radius        -0.325119             0.325119
1              mean texture        -0.494943             0.494943
2            mean perimeter        -0.317433             0.317433
3                 mean area        -0.395679             0.395679
4           mean smoothness        -0.188723             0.188723
5          mean compactness         0.436253             0.436253
6            mean concavity        -0.676390             0.676390
7       mean concave points        -0.843549             0.843549
8             mean symmetry        -0.334403             0.334403
9    mean fractal dimension         0.209551             0.209551
10             radius error        -1.376258             1.376258
11            texture error         0.045093             0.045093
12          perimeter error        -0.812926             0.812926
13               area error        -0.994695             0.994695
14        

**question 21 (1pt)**

Use the Pandas Dataframe `sort_values()` function to print the 5 features with the highest absolute value for its modelparameter. The first row should contain the feature with the highest absolute value.

In [23]:
#start code here

df_model_parameters.sort_values("parameter_value_abs",ascending=False).head(5)

#end code here

feature  parameter_value  parameter_value_abs
10          radius error        -1.376258             1.376258
21         worst texture        -1.058625             1.058625
13            area error        -0.994695             0.994695
27  worst concave points        -0.987129             0.987129
20          worst radius        -0.922528             0.922528